spark/src/main/java/org/apache/iceberg/spark/actions/BaseRepairManifestsSparkAction.java

```diff
@@ -0,0 +1,313 @@
+            Encoders.bean(ManifestFileInfo.class))
+        .toDF("partition_spec_id", "manifest_path");
+    Dataset<Row> manifestEntryDF = loadMetadataTable(getTable(), ENTRIES)
+        .filter("status < 2") // select only live entries
+        .selectExpr("input_file_name() as manifest_path", "snapshot_id", "sequence_number", "data_file");
+
+    return manifestEntryDF.as("manifest_entry")
+        .join(manifestDF.as("manifest"), "manifest_path")
+        .select("manifest_entry.manifest_path", "manifest.partition_spec_id",
+            "snapshot_id", "sequence_number", "data_file");
+  }
+
+  private static Function<Tuple2<ManifestFileInfo, Iterable<Row>>,
+      Tuple2<ManifestFileInfo, Iterable<PatchedManifestEntry>>> calculateRepairs(
+      Broadcast<Table> broadcastTable, Broadcast<SerializableConfiguration> conf,
+      StructType sparkType, RepairManifestHelper.RepairOptions options) {
+    return manifestFile -> {
+      Iterator<Row> rowIterator = manifestFile._2().iterator();
+      ManifestFileInfo manifestInfo = manifestFile._1();
+      int specId = manifestInfo.partSpecId;
+
+      PartitionSpec spec = broadcastTable.value().specs().get(specId);
+      Types.StructType dataFileType = DataFile.getType(spec.partitionType());
+      SparkDataFile wrapper = new SparkDataFile(dataFileType, sparkType).withSpecId(spec.specId());
+
+      Iterable<PatchedManifestEntry> entries =
+          Streams.stream(rowIterator).map(r -> {
+            long snapshotId = r.getLong(2);
+            long sequenceNumber = r.getLong(3);
+            Row file = r.getStruct(4);
+            SparkDataFile dataFile = wrapper.wrap(file);
+            Optional<DataFile> repairedDataFile = RepairManifestHelper.repairDataFile(
+                dataFile, broadcastTable.value(), spec, conf.value().value(), options);
+            return repairedDataFile.map(
+                value -> new PatchedManifestEntry(snapshotId, sequenceNumber, value, true))
+                .orElseGet(() -> new PatchedManifestEntry(snapshotId, sequenceNumber, dataFile, false));
+          }).collect(Collectors.toList());
+      return new Tuple2<>(manifestFile._1(), entries);
+    };
+  }
+
+  private static Function<Tuple2<ManifestFileInfo, Iterable<PatchedManifestEntry>>, ManifestFile> writeRepairs(
+      Broadcast<FileIO> io, Broadcast<Table> broadcastTable,
+      int formatVersion, String location) {
+    return rows -> {
+      Iterator<PatchedManifestEntry> entryIterator = rows._2().iterator();
+      ManifestFileInfo manifestInfo = rows._1();
+      int specId = manifestInfo.partSpecId;
+      PartitionSpec spec = broadcastTable.value().specs().get(specId);
+
+      String manifestName = "repaired-m-" + UUID.randomUUID();
+      Path newManifestPath = new Path(location, manifestName);
+      OutputFile outputFile = io.value().newOutputFile(FileFormat.AVRO.addExtension(newManifestPath.toString()));
```

Comment:
would it make sense to parallelize this as well? Using Tasks.foreach?

Reply:
Do you mean , parallelize this call: writer.existing(e.dataFile, e.snapshotId, e.sequenceNumber)?

It does not seem very multi-thread safe, looking through the writers in the writer stack (ManifestWriter, AvroFileAppender, DataFileWriter, GenericAvroWriter, the Avro writers..) have a lot of variables like counts that aren't safe.